Production OpenShift template for Jaeger

jpkrohling opened the pull request.

jpkrohling commented:
@pavolloffay , please review this PR, specially the readme

QE: this should be the state after this template is deployed:

```bash
$ oc get all
NAME                  REVISION   DESIRED   CURRENT   TRIGGERED BY
dc/jaeger-collector   1          1         1         config
dc/jaeger-query       1          1         1         config

NAME                    DESIRED   CURRENT   READY     AGE
rc/jaeger-collector-1   1         1         1         3m
rc/jaeger-query-1       1         1         1         3m

NAME                  HOST/PORT                                    PATH      SERVICES       PORT      TERMINATION   WILDCARD
routes/jaeger-query   jaeger-query-jaeger-infra.127.0.0.1.nip.io             jaeger-query   <all>     edge/Allow    None

NAME                   CLUSTER-IP       EXTERNAL-IP   PORT(S)                               AGE
svc/cassandra          None             <none>        7000/TCP,7001/TCP,7199/TCP,9042/TCP   3m
svc/jaeger-agent       None             <none>        5775/UDP,6831/UDP,6832/UDP            3m
svc/jaeger-collector   172.30.38.249    <none>        14267/TCP                             3m
svc/jaeger-query       172.30.193.126   <none>        80/TCP                                3m

NAME                     DESIRED   CURRENT   AGE
statefulsets/cassandra   2         2         3m

NAME                               DESIRED   SUCCESSFUL   AGE
jobs/jaeger-cassandra-schema-job   1         1            3m

NAME                                   READY     STATUS      RESTARTS   AGE
po/cassandra-0                         1/1       Running     0          3m
po/cassandra-1                         1/1       Running     0          2m
po/jaeger-agent-5d60v                  1/1       Running     0          3m
po/jaeger-cassandra-schema-job-8xrc3   0/1       Completed   0          3m
po/jaeger-collector-1-0mbpn            1/1       Running     0          3m
po/jaeger-query-1-grc80                1/1       Running     0          3m
```

![image](https://cloud.githubusercontent.com/assets/13387/26519473/be911f62-42c1-11e7-9623-0d18a0c59825.png)

For manual testing, the Hawkular APM's example can be used:
https://github.com/jpkrohling/hawkular-apm/tree/JPK-JaegerProduction . To see Traces on the UI, run the example as usual.

jpkrohling commented:
Test scenarios:

- [ ] Install the template into OpenShift and try to create Jaeger as an instant app (`oc create -f ...`)
- [ ] Create a Jaeger deployment by processing the template and passing it to the create procedure (`oc process -f ... | oc create -f -`)
- [ ] Undeploy
- [ ] Documentation testing (check the instructions for invalid steps/commands)

pavolloffay commented:
Great work :100: ! I will have a look on Monday.

pavolloffay commented:
Are you going to write some tests for it?

jpkrohling commented:
No, not at this iteration at least. QE could go one step further and write tests for it based on your `all-in-one` test, but I think it would be OK to not have tests on this iteration.

pavolloffay commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+    clusterIP: None
+    selector:
+      jaeger-infra: agent-instance
+parameters:
```
Could we move parameters to the top? It is cleaner what is required

jpkrohling replied:
Why cleaner? I think it's cleaner at the bottom :)

pavolloffay replied:
When you open the file you see all parameters at the top, it's like with a function/class/script, you also define parameters at the beginning. I don't know how people write  these templates maybe they do it like you did.

pavolloffay commented on README.md:
```diff
@@ -1,28 +1,157 @@
 
-## Troubleshooting
-Tracer sometimes fails to resolve Jaeger's address. In such a case run:
+To cleanup and remove everything, just run:
```
Actually there is a separete section, so this can be removed.

pavolloffay commented on README.md:
````diff
@@ -1,28 +1,157 @@
+oc delete all,template,daemonset -l jaeger-infra
+```
+
+## Production setup
````
This is a way too long readme with quite some repetitions. Could you please write it in a more compact and concise way? Please avoid a lot of text blocks like it is now.

Here should be a _short_ explanation what this template does, e.g. C* backend, that agent is one per node,  why it need system:admin user to run...

pavolloffay replied:
Could we add one root section with a hint how to access Jaeger's UI. the link should be similar to all-in-one..

jpkrohling replied:
@jstickler , is this something you could help us with? The challenge here is to have a minimum set of instructions where we would allow users to quickly try that out, but also providing enough information for an admin to properly install it in production.

jpkrohling replied:
Not the same, but similar.

JStickler replied:
I'm trying to get the release docs finished today.  I'll try to find some time to look at this in the next day or two.  At the very least, I can test out what you write and let you know if it's clear enough.

pavolloffay commented on README.md:
````diff
@@ -1,28 +1,157 @@
+```bash
+oc login -u system:admin
+oc new-project jaeger-infra
+oc process -f production/template.yml | oc create -f -
````
For example I would put here `-p NAMESPACE=jaeger-infra` and people would know what is going on, and the whole subsection can be omitted.

pavolloffay commented on README.md:
````diff
@@ -1,28 +1,157 @@
+oc policy add-role-to-user cluster-admin developer -n jaeger-infra
+```
+
+Your Agent hostname is `jaeger-agent.jaeger-infra.svc.cluster.local`
````
I guess this is visible only in the cluster? ping did not work.

jpkrohling replied:
Correct, for the agent to be available to the outside world, an external route has to be created for it.

pavolloffay commented on README.md:
```diff
@@ -1,28 +1,157 @@
+
+#### Post install
+
+The Docker images should be pulled automatically from Docker Hub, and all other processes should also complete
```
> The Docker images should be pulled automatically from Docker Hub

This is the way how templates usually work no? The most important thing from this long section is that some pods may fail for the first time.

pavolloffay commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+    triggers:
+    - type: ConfigChange
+- apiVersion: v1
+  kind: Service
```
It could contain a health check

pavolloffay commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+        restartPolicy: Always
+    triggers:
+    - type: ConfigChange
+- apiVersion: v1
```
it could define health check

pavolloffay commented:
It currently uses `:latest` could we use the same tag as for `all-in-one`?

pavolloffay commented on production/daemonset-admin.yml:
```diff
@@ -0,0 +1,31 @@
+#
```
Could we change a name to be more specific?

jpkrohling replied:
Suggestion?

pavolloffay replied:
Is it related to the production template? Then `jaeger-production-daemonset-admin.yml` If not without production?

jpkrohling replied:
Not really, this is a generic role, which can be added to users, so that said user can create daemon sets. Not really sure this will be used in a real world environment, where a cluster-admin would deploy Jaeger.

pavolloffay commented:
If somebody is using our openshift online, can he connect as `system:admin` if not we should provide a way to deploy the production version there too.

jpkrohling commented on README.md:
````diff
@@ -1,28 +1,157 @@
+to be issued:
+
+```bash
+oc login -u system:admin
````
This is inside a section named "... as admin". There's another one, for "... as developer". 
For the Agent to work as a Daemon Set, it does need to be installed as admin (or at least, privileged user), as the daemon set runs at the node level.

Perhaps we could provide another template, for environments like openshift.io , where the Agent won't run as a daemon set?

jpkrohling commented on README.md:
````diff
@@ -1,28 +1,157 @@
+```bash
+oc login -u system:admin
+oc new-project jaeger
+oc process -f template.yml | oc create -f -
````
Actually, the `oc new-project` should create `jaeger-infra`, not `jaeger`. It seems there's a convention that non-business namespaces are named with a `-infra` suffix.

jpkrohling commented on README.md:
````diff
@@ -1,28 +1,157 @@
+jaeger-query-1-h089j                1/1       Running     3          55s
+```
+
+## Getting an OpenShift cluster running
````
I would keep at least a very basic set of instructions. OpenShift is a complex software, and there are literally thousands of ways to deploy it. Mentioning `oc cluster up` and `minishift` should be short enough, while still pointing users to the "easiest" solutions for a quick start.

jpkrohling commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+  name: jaeger-template
+  annotations:
+    description: Jaeger Distributed Tracing Server
+    iconClass: icon-cogs
```
http://fontawesome.io/3.2.1/icon/cogs/

Ideally, we should get an icon-jaeger ;) Is this something @khala could provide?

pavolloffay replied:
I think the icon has to be installed on openshift. There is already https://www.todobackend.com/images/logos/golang.png (all-in-one) is using it. We could use that as an intermediate solution.

jpkrohling commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+      jaeger-infra: cassandra-statefulset
+  spec:
+    serviceName: cassandra
+    replicas: 2
```
Because it's more than 1 :) We could add 1 node, but this is to "prove" that users can scale Cassandra.

jpkrohling commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+      spec:
+        containers:
+        - name: cassandra
+          image: jpkroehling/cassandra
```
We discussed that at the beginning, IIRC. This is meant to be temporary, as we have another task about using the same Cassandra for all Hawkular components. I added a issue, to get this sorted out.

jpkrohling commented on production/template.yml:
```diff
@@ -0,0 +1,321 @@
+        - name: cassandra
+          image: jpkroehling/cassandra
+          ports:
+          - containerPort: 7000
```
You mean the port list? I need to read the docs on this, I'm also not quite happy with that.

jpkrohling commented:
For `openshift.io`, I'd create yet another template. The proper way on bigger deployments is to have one agent per node, and a daemon set is the appropriate way for that. For individual deployments, like in openshift.io , the end user won't be able to deploy a daemon set (I believe).

jpkrohling commented:
PR updated with the fixes from the review. @pavolloffay , would you please re-review? Some changes will be done as part of the issues I created (like the Cassandra one, the OpenShift.io template and so on).

pavolloffay commented:
Sure, one thing could you please use tagged versions of the images? Perhaps the same as all-in-one

jpkrohling commented:
> Sure, one thing could you please use tagged versions of the images? Perhaps the same as all-in-one

Why? If you set it to pull only if not present, the administrator is able to control it. If you set a tag, you force the admin to change your template if they want/need to use a specific version.

pavolloffay commented:
> Why? If you set it to pull only if not present, the administrator is able to control it. If you set a tag, you force the admin to change your template if they want/need to use a specific version.

Isn't this something we want to control? e.g. say that this template has been tested against jaeger version `xxx` which is in the template? If they want to use a different version it is up to them, they can still change the template.

jpkrohling commented:
No, not at this point at least. There are several ways to view this, and we might not know what our users want at this point. Let's do a "continuous delivery", and if we are breaking stuff too often, then we set a tag.

pavolloffay commented:
I think here we should provide always working solution, so when users run it always run without any problems. I see benefits of `IfNotPresent` when the admin controls templates and what local registry which is not the case of public templates. 

We are already versioning this repo and all-in-one is associated with a tagged image. For instance it is used in MSA example, without a tag we cannot guarantee that it will _always_ work. 

Does it make sense?

pavolloffay commented:
@objectiser  ^^ do you have any preference?

objectiser commented:
Would suggest getting input from obsidian team.

jpkrohling commented:
Note that nothing is final: on special cases, the admin can edit the DC/RC to set their own images and/or versions. For now, this first release, where we don't know how people are going to consume it, let's keep it simple and in line to what is done out there.

pavolloffay commented:
> Would suggest getting input from obsidian team.

@objectiser I don't think this is related to obsidian. But fabric8 provide resources with tagged versions. 

> let's keep it simple and in line to what is done out there.

Could you point me to any public repo providing OC templates which is using `IfNotPresent` strategy? So far I have seen only tagged versions :/.

pavolloffay commented on production/jaeger-production-template.yml:
```diff
@@ -0,0 +1,327 @@
+      jaeger-infra: cassandra-statefulset
+  spec:
+    serviceName: cassandra
+    replicas: 2
```
What is the advantage of 2 replicas? Can users scale this up when necessary?

Do you know memory footprint of one c* pod?

yurishkuro replied:
note that it doesn't mean a primary + 2 replicas, it's a total of 2 copies of data in the cluster (there are no primaries in Cassandra)

jpkrohling replied:
Correct, `replica` here is an OpenShift term, and has nothing to do with the `replica` from Cassandra. And indeed, the number of replicas here is to indicate that this is what we "recommend" for a production usage, although an admin will certainly adjust it based on the target deployment scenario.

pavolloffay commented:
Tu document I have sent an email to http://post-office.corp.redhat.com/archives/obsidian-team-list/2017-May/msg00333.html (it might be accessible only in RH domain)

jpkrohling commented:
PR updated. I changed the versions to be what is recommended by the Kubernetes documents, assuming that the SHA is an actual version tag. We can later change the templates to use `latest`, if that's what we want.

I also updated the `all-in-one`, to align with the production one (icon, descriptions, urls, namespace, ...)

I believe there are no more remaining tasks for this PR, as most of the other concerns are on other issues (readme, cassandra, ...), but I'd appreciate a new review :)

jkandasa commented:
@jpkrohling @pavolloffay @objectiser I tried on `minihift` and `oc cluster up`. Failed on both targets.

Version information:
`minishift`
```
OpenShift Master: v1.5.0+031cbe4
Kubernetes Master: v1.5.2+43a9be4
```
`oc cluster up`
```
OpenShift Master: v1.5.1+7b451fc
Kubernetes Master: v1.5.2+43a9be4
```
##### Issue on `all-in-one/jaeger-all-in-one-template.yml`
![all-in-one](https://cloud.githubusercontent.com/assets/1004403/26676943/612ca646-46e7-11e7-932a-be0767983098.png)


##### Issue on `production/jaeger-production-template.yml`
![jaeger-production-template](https://cloud.githubusercontent.com/assets/1004403/26676954/6e9e6c4c-46e7-11e7-8240-a2215821e82f.png)

jkandasa commented on README.md:
````diff
@@ -1,28 +1,157 @@
+```bash
+oc login -u system:admin
+oc create -f daemonset-admin.yml
+oadm policy add-role-to-user daemonset-admin developer
````
I guess we have to replace `oadm` as `oc`?

jpkrohling replied:
I'll keep this as `oadm` for the moment, as the switch from `oadm` to `oc adm` is relatively recent, so, it might confuse people using older `oc` versions. But that's a good catch!

jkandasa commented:
@jpkrohling When I follow instructions from [README](https://github.com/jpkrohling/jaeger-openshift/blob/cf3a09521afa21d7b492a9d50a6c3d0106fbb756/README.md#installing-jaeger-under-an-existing-namespace-as-developer) file for `developer` deployed successfully via command line. However, it fails on GUI as ` in the screenshot.

jpkrohling commented:
@jkandasa , on the GUI, the `developer` user has to have the daemon-set role (see the README), and you should either adapt the `namespace` var to be `my-project`, or create a new project `jaeger-infra` and create the resources there.

I could not reproduce the second error, though.

jkandasa commented:
@jpkrohling As shown in README file, I have executed the following commands to get required privilege for `developer` user.

```
oc login -u system:admin
oc create -f daemonset-admin.yml
oc policy add-role-to-user daemonset-admin developer
```

Am I missing anything?

But the error says like syntax error?
`oc cluster up`
```
OpenShift Master: v1.5.1+7b451fc
Kubernetes Master: v1.5.2+43a9be4
```
![jaeger-production-template](https://cloud.githubusercontent.com/assets/1004403/26676954/6e9e6c4c-46e7-11e7-8240-a2215821e82f.png)

jpkrohling commented:
I reproduced the second issue. This happens only when importing the template via the UI. I'll open an issue on the OpenShift project.

jpkrohling commented:
But the first problem sounds like you missed some step.

From the `production` directory:
```
$ oc cluster up --version=latest
$ oc login -u system:admin
$ oc create -f daemonset-admin.yml
$ oc policy add-role-to-user daemonset-admin developer
$ oc create -f jaeger-production-template.yml
```

Then, on the UI, change the namespace, to be the same as the project you currently are:

![image](https://cloud.githubusercontent.com/assets/13387/26679659/6608dfc4-46d6-11e7-9017-2327c2b459d5.png)

Results in:

![image](https://cloud.githubusercontent.com/assets/13387/26679676/757b62f6-46d6-11e7-99b9-03328018915f.png)

jkandasa commented:
@jpkrohling all works. Thank you! Now I will do validation part.

pavolloffay commented on README.md:
```diff
@@ -1,28 +1,157 @@
+
+Your Agent hostname is `jaeger-agent.jaeger-infra.svc.cluster.local`
+
+Clean up with:
```
Redundant there is a separate section which applies for all templates

jpkrohling replied:
Should be done in #9

pavolloffay commented on README.md:
```diff
@@ -1,28 +1,157 @@
 
-## Troubleshooting
-Tracer sometimes fails to resolve Jaeger's address. In such a case run:
+Once everything is ready, the Jaeger Query UI can be accessed via https://jaeger-query-jaeger.127.0.0.1.nip.io/ .
```
Could you change the link to `https://jaeger-query.127.0.0.1.nip.io/` ?

pavolloffay replied:
This worked for me `http://jaeger-jaeger-infra.10.40.1.51.xip.io/`

pavolloffay commented on README.md:
```diff
@@ -1,28 +1,157 @@
+
+## Production setup
+
+### TL;DR
```
It would be good to write a few high level words about the template, e.g. it's meant for OC on premise, uses C* agent is deployed per node and therefore requires admin rights.

jpkrohling replied:
Further changes to the readme will be done in the context of #9 .

pavolloffay commented on README.md:
```diff
@@ -1,28 +1,157 @@
+
+Once everything is ready, the Jaeger Query UI can be accessed via https://jaeger-query-jaeger.127.0.0.1.nip.io/ .
+
+Your Agent hostname is `jaeger-agent.jaeger-infra.svc.cluster.local`
```
How this can be verified? Ping does not work?

jpkrohling replied:
Ping works:
```
oc exec -it cassandra-0 bash
bash-4.2$ ping jaeger-agent.jaeger-infra.svc.cluster.local
PING jaeger-agent.jaeger-infra.svc.cluster.local (172.17.0.4) 56(84) bytes of data.
64 bytes from 172.17.0.4 (172.17.0.4): icmp_seq=1 ttl=64 time=0.181 ms
64 bytes from 172.17.0.4 (172.17.0.4): icmp_seq=2 ttl=64 time=0.919 ms
```

It can be verified by using an application that reports spans to the agent, like this:
https://github.com/jpkrohling/hawkular-apm/tree/JPK-JaegerProduction

pavolloffay commented on production/jaeger-production-template.yml:
```diff
@@ -0,0 +1,345 @@
+          image: jaegertracing/jaeger-agent:${IMAGE_VERSION}
+          command:
+          - "/go/bin/agent-linux"
+          - "-collector.host-port=jaeger-collector:14267"
```
jaeger-collector shouldn't this be a variable?

jpkrohling replied:
Why should it? `jaeger-collector` is not a real host, it is the Kubernetes service name, which is the official discovery mechanism in Kubernetes.

pavolloffay commented on production/jaeger-production-template.yml:
```diff
@@ -0,0 +1,345 @@
+- apiVersion: v1
+  kind: Service
+  metadata:
+    namespace: "${NAMESPACE}"
```
I'm maybe too much pedantic, but quotes are not necessary here also in apiVersions and other definitions

pavolloffay replied:
Just saying because you are not using them always.

pavolloffay commented on production/jaeger-production-template.yml:
```diff
@@ -0,0 +1,345 @@
+  app: jaeger
+  template: jaeger-template
+  jaeger-infra: template
+message: 'A Jaeger service has been created in your project.'
```
are quotes necessary? They are not used above

pavolloffay commented on all-in-one/jaeger-all-in-one-template.yml:
```diff
@@ -12,52 +12,63 @@
+  value: jaeger
+- description: The namespace where to install the Jaeger instance
+  displayName: Namespace
+  name: NAMESPACE
```
Could we make this optional? By default install it in the current namespace? It should be possible with prod template too, then the instructions would be more specific and clear from beginning that it will use jaeger-infra namespace.

jpkrohling replied:
This is actually not required at all. It's there because we wanted to align it with the production template. And the production template has it set to `jaeger-infra` because it's actually a good idea to have a namespace dedicated to Jaeger and it should be our recommendation.

pavolloffay replied:
One thing, can apps see each other in different namespaces?

jpkrohling replied:
Yes

pavolloffay commented on README.md:
````diff
@@ -1,28 +1,157 @@
+
+Once that is ready, it's only a matter of creating the components from the template:
+```bash
+oc process -f jaeger-production-template.yml -p NAMESPACE=myproject | oc create -f -
````
Maybe a note that myproject has to be created before issuing `oc create -f deamon`. 

Or actually `oc create -f daemonset-admin.yml` is maybe applied to a current namespace so maybe `oc create -f daemonset-admin.yml --namespace=myproject` would help

jpkrohling replied:
To be taken care of on #9

jkandasa commented:
@jpkrohling @pavolloffay Tested deployment and cleanup for the templates `jaeger-production-template.yml` and `jaeger-all-in-one-template.yml`. both templates deployed as expected. I just launched UI of Jaeger works well. Not tested with `span`'s.

delete also works as expected. Command: `oc delete all,template,daemonset -l jaeger-infra`

pavolloffay commented:
Merging thanks @jpkrohling @jkandasa

jpkrohling commented:
![](https://github.trello.services/images/mini-trello-icon.png) [UC-1: Openshift Deployment](https://trello.com/c/M3oVmpPR/1-uc-1-openshift-deployment)
